File: 0139.word_break/solution.py
from typing import List


class Solution:
    def word_break(self, s: str, word_dict: List[str]) -> bool:
        length = len(s)
        dp = [False] * (length + 1)
        dp[0] = True
        word_dict = set(word_dict)
        for i in range(1, length+1):
            for j in range(i):
                if dp[j] and s[j:i] in word_dict:
                    dp[i] = True
                    break
        return dp[-1]

File: 0139.word_break/solution.go
package main

func wordBreak(s string, wordDict []string) bool {
	b := []byte(s)
	dp := make([]bool, len(b))
	for i := 0; i < len(s); i++ {
		for _, key := range wordDict {
			if i < len(key)-1 {
				continue
			}
			if (i == len(key)-1 || dp[i-len(key)]) && string(b[i-len(key)+1:i+1]) == key {
				dp[i] = true
			}
		}
	}
	return dp[len(s)-1]
}
